src/main/java/com/linecorp/armeria/client/limit/ConcurrencyLimitingClient.java

```diff
@@ -0,0 +1,232 @@
+                                        int maxConcurrency, long timeout, TimeUnit unit) {
+        super(delegate);
+
+        validateAll(maxConcurrency, timeout, unit);
+
+        this.maxConcurrency = maxConcurrency;
+        timeoutMillis = unit.toMillis(timeout);
+    }
+
+    static void validateAll(int maxConcurrency, long timeout, TimeUnit unit) {
+        validateMaxConcurrency(maxConcurrency);
+        if (timeout < 0) {
+            throw new IllegalArgumentException("timeout: " + timeout + " (expected: >= 0)");
+        }
+        requireNonNull(unit, "unit");
+    }
+
+    static void validateMaxConcurrency(int maxConcurrency) {
+        if (maxConcurrency < 0) {
+            throw new IllegalArgumentException("maxConcurrency: " + maxConcurrency + " (expected: >= 0)");
+        }
+    }
+
+    /**
+     * Returns the number of the {@link Request}s that are being executed.
+     */
+    public int numActiveRequests() {
+        return numActiveRequests.get();
+    }
+
+    @Override
+    public O execute(ClientRequestContext ctx, I req) throws Exception {
+        return maxConcurrency == 0 ? unlimitedExecute(ctx, req)
+                                   : limitedExecute(ctx, req);
+    }
+
+    private O limitedExecute(ClientRequestContext ctx, I req) throws Exception {
+        final Deferred<O> deferred = defer(ctx, req);
+        final PendingTask currentTask = new PendingTask(ctx, req, deferred);
+
+        pendingRequests.add(currentTask);
+        drain();
+
+        if (!currentTask.isRun() && timeoutMillis != 0) {
+            // Current request was not delegated. Schedule a timeout.
+            final ScheduledFuture<?> timeoutFuture = ctx.eventLoop().schedule(
+                    () -> deferred.close(ResponseTimeoutException.get()),
+                    timeoutMillis, TimeUnit.MILLISECONDS);
+            currentTask.set(timeoutFuture);
+        }
+
+        return deferred.response();
+    }
+
+    private O unlimitedExecute(ClientRequestContext ctx, I req) throws Exception {
+        numActiveRequests.incrementAndGet();
+        boolean success = false;
+        try {
+            final O res = delegate().execute(ctx, req);
+            res.closeFuture().whenComplete((unused, cause) -> numActiveRequests.decrementAndGet());
+            success = true;
+            return res;
+        } finally {
+            if (!success) {
+                numActiveRequests.decrementAndGet();
+            }
+        }
+    }
+
+    void drain() {
+        for (;;) {
+            final int currentActiveRequests = numActiveRequests.get();
+            if (currentActiveRequests >= maxConcurrency) {
+                break;
+            }
+
+            if (numActiveRequests.compareAndSet(currentActiveRequests, currentActiveRequests + 1)) {
+                final PendingTask task = pendingRequests.poll();
+                if (task == null) {
+                    numActiveRequests.decrementAndGet();
```

Comment:
> If there is new request comes in at another thread, it will jump out at LN 138?
Yes. However, once the first request is complete [here](https://github.com/line/armeria/pull/233/files#diff-a3df543be87f334befaccfbf2fa6b383R221), `decrementAndGet()` and `drain()` will be called to catch up what's in the queue.